Rewrite DELETE docs (#55856)
ClickHouse/ClickHouse: docs/en/sql-reference/statements/delete.md

title: DELETE Statement
---

The `DELETE` statement removes rows from the table `[db.]table` that match the expression `expr`.

``` sql
DELETE FROM [db.]table [ON CLUSTER cluster] WHERE expr;
```

> Deleted rows are internally marked as deleted immediately and will be automatically filtered out of all subsequent queries. Cleanup of data happens asynchronously in the background until the next merge. As a result, it is possible that for an unspecified period, data is not actually deleted from storage and is only marked as deleted.

Reviewer: This would be more precise
```suggestion
Deleted rows are internally marked as deleted immediately and will be automatically filtered out of all subsequent queries. Cleanup of data happens during the next merge. As a result, it is possible that for an unspecified period, data is not actually deleted from storage and is only marked as deleted.
```

> 

Reviewer: Should we also somewhere at the top of the page mention and refer people to a "heavy" `ALTER table DELETE` query that will actually give you guarantees (asynchronously or synchronously) of deleting your data in a predictable period of time?

Deleted rows are internally marked as deleted immediately and will be automatically filtered out of all subsequent queries. Cleanup of data happens asynchronously in the background until the next merge. As a result, it is possible that for an unspecified period, data is not actually deleted from storage and is only marked as deleted.

The `DELETE` statement is only available for the *MergeTree table engine family.

For example, the following query deletes all rows from the `hits` table where the `Title` column contains the text `hello`:
**Do not use the DELETE statement frequently as it can negatively affect ClickHouse performance.**

## Examples

```sql
-- Deletes all rows from the `hits` table where the `Title` column contains the text `hello`
DELETE FROM hits WHERE Title LIKE '%hello%';
```

> Lightweight deletes are asynchronous by default. Set `mutations_sync` equal to 1 to wait for one replica to process the statement, and set `mutations_sync` to 2 to wait for all replicas.

Reviewer: Why were all mentions of `DELETE` being lightweight deleted?

Author: Sorry, I'm not sure I understand this comment. Can you expand further?

Reviewer: I mean previously DELETEs in the text were explicitly called lightweight, now it is just DELETEs (except line 75)

title: DELETE Statement
---

The `DELETE` statement removes rows from the table `[db.]table` that match the expression `expr`.

``` sql
DELETE FROM [db.]table [ON CLUSTER cluster] WHERE expr;
```

> Deleted rows are internally marked as deleted immediately and will be automatically filtered out of all subsequent queries. Cleanup of data happens asynchronously in the background. The `DELETE` statement is only available for the *MergeTree table engine family.

Reviewer: Should we specifically say that this is a "lazy" type of deletion and not that the deletion will happen asynchronously in the background, because we will not do anything to actively delete the data but will simply wait until the next merge happens? As a result, the deleted data can be stored for a long time without actually being deleted.

> **Do not use the DELETE statement frequently as it can negatively affect ClickHouse performance.**

Reviewer: I am also not sure about this statement because yes, it won't really affect the performance as much as it will just take an undefined amount of time to delete the data.

Author: What's a better way to phrase this?

**Deleting large volumes of data frequently can negatively affect ClickHouse performance.**

Or should we just omit this?

Reviewer: Maybe something like this?
```
Deleting large volumes of data with the DELETE statement can negatively affect SELECT query performance.
```

## Examples

```sql
-- Deletes all rows from the `hits` table where the `Title` column contains the text `hello`
DELETE FROM hits WHERE Title LIKE '%hello%';
```

## Handling large `DELETE`s

> Large deletes can negatively affect ClickHouse performance. If you are attempting to delete all rows from a table, consider using the [`TRUNCATE TABLE`](/en/sql-reference/statements/truncate) command and then re-create the table.

Reviewer: What does `re-create the table` mean in this context? TRUNCATE TABLE deletes all the data in the table but it does not DROP it, so there isn't really anything to recreate

> If you anticipate frequent deletes, consider using a [customer partitioning key](/en/engines/table-engines/mergetree-family/custom-partitioning-key). You can then use the [`ALTER TABLE...DROP PARTITION`](/en/sql-reference/statements/alter/partition#drop-partitionpart) command to quickly drop all rows associated with that partition.

Reviewer: ```suggestion
If you anticipate frequent deletes, consider using a [custom partitioning key](/en/engines/table-engines/mergetree-family/custom-partitioning-key). You can then use the [`ALTER TABLE...DROP PARTITION`](/en/sql-reference/statements/alter/partition#drop-partitionpart) command to quickly drop all rows associated with that partition.
```

## Limitations of `DELETE`

### `DELETE`s are eventually consistent

> `DELETE`s are asynchronous by default. This means that if you use `DELETE` in a multi-replica setup, your rows will not be deleted immediately. To change this behavior, set `mutations_sync` equal to 1 to wait for one replica to process the statement, or set `mutations_sync` to 2 to wait for all replicas.

Reviewer: This is not true, the DELETE (marking the rows as deleted) is synchronous by default (mutations_sync = 2), and there is currently nothing except `OPTIMIZE TABLE ... FINAL` that can force ClickHouse to actually synchronously delete the data.
